src/queue.rs

```diff
@@ -660,46 +649,52 @@ impl<M: GuestAddressSpace> Queue<M> {
+    fn used_event(&self, order: Ordering) -> Result<Wrapping<u16>, Error> {
         // Safe because we have validated the queue and access guest memory through GuestMemory
         // interfaces.
-        // And the `used_index` is a two-byte naturally aligned field, so it won't cross the region
-        // boundary and get_slice() shouldn't fail.
         let mem = self.mem.memory();
         let used_event_addr = self
             .avail_ring
             .unchecked_add((4 + self.actual_size() * 2) as u64);
 
-        // Complete all the writes in add_used() before reading the event.
-        fence(Ordering::SeqCst);
-        mem.get_slice(used_event_addr, size_of::<u16>())
-            .map(|s| {
-                Wrapping(
-                    s.get_atomic_ref::<AtomicU16>(0)
-                        .unwrap()
-                        .load(Ordering::Relaxed),
-                )
-            })
-            .map_err(|_| Error::GuestMemoryError)
+        mem.load(used_event_addr, order)
+            .map(Wrapping)
+            .map_err(Error::GuestMemory)
     }
 
     /// Check whether a notification to the guest is needed.
-    pub fn needs_notification(&mut self, used_idx: Wrapping<u16>) -> bool {
-        let mut notify = true;
+    ///
+    /// Please note this method has side effects: once it returns `true`, it considers the
+    /// driver will actually be notified, remember the associated index in the used ring, and
+    /// won't return `true` again until the driver updates `used_event` and/or the notification
+    /// conditions hold once more.
+    pub fn needs_notification(&mut self) -> Result<bool, Error> {
+        let used_idx = self.next_used - Wrapping(1);
+
+        // Complete all the writes in add_used() before reading the event.
+        fence(Ordering::SeqCst);
 
         // The VRING_AVAIL_F_NO_INTERRUPT flag isn't supported yet.
         if self.event_idx_enabled {
-            if let Some(old_idx) = self.signalled_used.replace(used_idx) {
-                // Silently ignore errors and let the caller send the notification.
-                if let Ok(used_event) = self.used_event() {
-                    if (used_idx - used_event - Wrapping(1u16)) >= (used_idx - old_idx) {
-                        notify = false;
-                    }
+            if let Some(prev) = self.signalled_used {
+                let used_event = self.used_event(Ordering::Relaxed)?;
+                // This check looks at `used_idx`, `used_event`, and `prev` as if they are on an
+                // axis (that wraps around). If the difference between `used_idx` and `used_used`
+                // is greater than or equal to the difference between `used_idx` and `prev`, then
+                // `prev` is closer to `used_idx` than `used_event` (and thus more recent), so
+                // we don't need to elicit another notification.
+                if used_idx - used_event >= used_idx - prev {
```

Comment:
Why the "- Wrapping(1u16)" has been removed?

Reply:
It hasn't been removed entirely, it's now part of the computation at the beginning of the function (line 672) which computes the last index in the used ring where we wrote something.

Reply:
Seems it still has changed the logic because the `prev` is self.next_used - Wrapping(1) of last round?

Reply:
Oh, I see your point now. So essentially, the left-hand side of the inequality is the same as before (because `used_event` is the same, and `used_idx` is now defined as `self.next_used - Wrapping(1)`), but the right-hand side is different because `used_idx` is now the old `used_idx` with a `- Wrapping(1)` as well. After a bit of doodling I think you are right and the change is erroneous. We should use a `>` instead of `>=` to account for the implicit `- Wrapping(1)` so things are exactly as before.

This also has the added advantage that `needs_notification` will return true until `used_idx` actually changes, and I think we can use a value that defaults to 0 for `signalled_used` instead of an `Option`. I'll experiment a bit more with the changes and update the PR later today or tomorrow at the latest. Thanks for pointing this out!

Reply:
Be careful about the ">=" vs ">" and wrapping around issue:)
All implementation of algorithm I have seen using the form "if (used_idx - used_event - Wrapping(1u16)) >= (used_idx - old_idx)".

Reply:
Ok so after more playing around, the sad conclusion is there's no nice way of making `needs_notification` idempotent. I think a cleaner solution is something like making `add_used` return a boolean value that indicates whether a notification is needed or not. For the time being, I've changed the code here to be essentially the same as before the  PR, for ease of re-reviewing this part.

Also, turns out the regression caused by the original version of the PR was not around the handling of `- Wrapping(1)` (which actually worked out in the end), but the fact that `signalled_used` was only updated when the method would return `true` (which fit the name, but changed the logic).